tests/kernel/mutex/mutex_api/src/test_mutex_apis.c

```diff
@@ -140,18 +236,171 @@ void test_mutex_lock_unlock(void)
 	tmutex_test_lock_unlock(&kmutex);
 }
 
+/**
+ * @brief Test recursive mutex
+ * @details To verify that getting a lock of a mutex already locked will
+ * succeed and waiters will be unblocked only when the number of locks
+ * reaches zero.
+ * @ingroup kernel_mutex_tests
+ */
+void test_mutex_recursive(void)
+{
+	k_mutex_init(&mutex);
+
+	/**TESTPOINT: when mutex has no owner, we cannot unlock it */
+	zassert_true(k_mutex_unlock(&mutex) == -EINVAL,
+			"fail: mutex has no owner");
+
+	zassert_true(k_mutex_lock(&mutex, K_NO_WAIT) == 0,
+			"Failed to lock mutex");
+
+	/**TESTPOINT: lock the mutex recursively */
+	zassert_true(k_mutex_lock(&mutex, K_NO_WAIT) == 0,
+		"Failed to recursively lock mutex");
+
+	/* Spawn a waiter thread */
+	k_thread_create(&tdata3, tstack3, STACK_SIZE,
+			(k_thread_entry_t)tThread_waiter, &mutex, NULL, NULL,
+			K_PRIO_PREEMPT(12),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	zassert_true(thread_ret == TC_FAIL,
+		"waiter thread should block on the recursively locked mutex");
+
+	zassert_true(k_mutex_unlock(&mutex) == 0, "fail to unlock");
+
+	/**TESTPOINT: unlock the mutex recursively */
+	zassert_true(thread_ret == TC_FAIL,
+		"waiter thread should still block on the locked mutex");
+
+	zassert_true(k_mutex_unlock(&mutex) == 0, "fail to unlock");
+
+	/* Give thread_waiter a chance to get the mutex */
+	k_sleep(K_MSEC(1));
+
+	/**TESTPOINT: waiter thread got the mutex */
+	zassert_true(thread_ret == TC_PASS,
+			"waiter thread can't take the mutex");
+}
+
+/**
+ * @brief Test mutex's priority inheritance mechanism
+ * @details To verify mutex provide priority inheritance to prevent prority
+ * inversion, and there are 3 cases need to run.
+ * The thread T1 hold the mutex first and cases list as below:
+ * - case 1. When prority T2 > T1, priority inheritance happened.
```

Comment:
Can you make an enum for the three cases?

Reply:
@sjg20 ok, I'll make an enum for case 1, 2, 3 in next submit. thanks for reminding that!